Adding murmurhash3 digest to phobos

gchatelet opened the pull request:
A proposal of addition of murmurhash3 to phobos.
Maybe it should go in experimental first. Please let me know what you think.

9il commented on std/digest/murmurhash3.d:
```diff
@@ -0,0 +1,734 @@
+
+public:
+
+    void putBlocks(const(Block)[] blocks) pure nothrow @nogc
```
Could you replace `Block` with `int`. `Block` is private.

9il commented on std/digest/murmurhash3.d:
```diff
@@ -0,0 +1,734 @@
+
+public:
+
+    void putBlocks(const(Block)[] blocks) pure nothrow @nogc
```
ditto (about block)

9il commented on std/digest/murmurhash3.d:
```diff
@@ -0,0 +1,734 @@
+    uint h4, h3, h2, h1;
+    size_t size;
+
+    this(uint seed4, uint seed3, uint seed2, uint seed1)
```
Should it be public?

9il commented on std/digest/murmurhash3.d:
```diff
@@ -0,0 +1,734 @@
+    ulong h2, h1;
+    size_t size;
+
+    this(ulong seed)
```
ditto (public?)

9il commented on std/digest/murmurhash3.d:
```diff
@@ -0,0 +1,734 @@
+
+public:
+
+    void putBlocks(const(Block)[] blocks) pure nothrow @nogc
```
ditto (Block)

9il commented on std/digest/murmurhash3.d:
```diff
@@ -0,0 +1,734 @@
+        return buffer[0 .. bufferSize];
+    }
+
+    size_t alignDownTo(size_t size) const
```
static

9il commented on std/digest/murmurhash3.d:
```diff
@@ -0,0 +1,734 @@
+        }
+        // Pushing as many consecutive blocks as possible.
+        immutable consecutiveBlocks = alignDownTo(data.length);
+        hasher.putBlocks(cast(const(Block)[]) data[0 .. consecutiveBlocks]);
```
Is it safe? How it would work if data is not aligned? (may be create a small scope buffer for tail and and unaligned head?)

gchatelet commented:
Thx for the comments @9il.
I still need to spend some time on the documentation.
I'll add some tests to check if there's an alignment issue when pushing consecutive blocks in the Piecewise struct.

andralex commented:
LGTM from 30K feet. Any experts on board to vet this?

9il commented:
> LGTM from 30K feet. Any experts on board to vet this?

MurmurHash implementation looks good. `SMurmurHash3_XXX_XXx` would be very useful for fast user defined hash functions (especially for containers or `ndslice`, where strides between elements can be greater than 1).

9il commented on std/digest/murmurhash3.d:
```diff
@@ -0,0 +1,760 @@
+    const(ubyte)[] getRemainder()
+    {
+        return buffer[0 .. bufferSize];
+    }
```
Probably all private functions should be marked with `pragma(inline, true):`

gchatelet commented:
@9il I added the test about alignement you were worried about.
I still need to add some documentation.

jpf91 commented:
>  I added the test about alignement you were worried about.

Your test/code works on X86 but it won't work on architectures which do not properly support [unaligned loads](https://www.kernel.org/doc/Documentation/unaligned-memory-access.txt). This includes older ARM devices so I'd say this is a blocker. Instead of this code:

```d
hasher.putBlocks(cast(const(Block)[]) data[0 .. consecutiveBlocks]);
```

you'll have to copy n bytes into a block/ubyte union:
```d
union
{
    Block block;
    ubyte[Block.sizeof] rawData;
} buf;

foreach(...)
{
    buf.rawData[] = data[n..n+Block.sizeof];
    hasher.putBlock(...);
}
```

If benchmarking shows that the copying makes the algorithm much slower you could do a runtime check whether data.ptr is properly aligned and then skip the copying.

gchatelet commented:
Thx @jpf91 this is very helpful. I'll test different versions but I expect the performance to drop dramatically. According to [the original C++ implementation](https://code.google.com/p/smhasher/wiki/MurmurHash3): _that's about 1.68 bytes per cycle, or about 9.5 cycles per 16-byte chunk._ this allows to achieve 5GB/sec for MurmurHash3_x64_128.
The current implementation is not as fast but I achieved 4.3GB/sec (different computer so not necessarily comparable). I'll report here my findings but copying is not free so performance will suffer.

On an architecture note, the implementation is currently suboptimal for big-endian.
See comment on endianness on [the wikipedia page](https://en.wikipedia.org/wiki/MurmurHash#Algorithm). I need to fix this also.

9il commented:
> I'll test different versions but I expect the performance to drop dramatically.

You can use `version(X86)` and others to determine if algorithm can use unaligned loads.

jpf91 commented:
> You can use version(X86) and others to determine if algorithm can use unaligned loads.

That certainly should be done. It would still be nice to have a fast-path for aligned data on architectures without unaligned loads as well. Here are some possible solutions:

1. Your `put` method isn't marked as `safe` or `trusted`. So you could probably argue that your function isn't memory safe and document that the programmer is responsible for passing only aligned data.

2. You can check `data.ptr`  and select the codepath for aligned or unaligned data at runtime. This will probably still have some performance impact (depending on the user code / data size passed to `put`). But you could combine this approach with `version` statements.

3. You could also add an `put(const(uint)[])` overload. All data passed to this overload would be guaranteed to be aligned. But:
 * I'm not sure how well this actually works with range composition.
 * It doesn't work that well for the object-oriented interface. You could add it as an additional method but then it still couldn't be used with generic code using the `Digest` interface. Adding a new method to the interface could break user code. As it's safe to cast `uint[]` to `ubyte[]` the overload for aligned data could always pass the data to the overload for unaligned data. So we'd have to make `Digest` an abstract class instead of an interface and provide a default implementation.

Pinging some people who have worked on `std.digest` for more opinions: @schuetzm @CyberShadow @klickverbot

schuetzm commented:
> 1. Your put method isn't marked as safe or trusted. So you could probably argue that your function isn't memory safe and document that the programmer is responsible for passing only aligned data.

I don't think `@safe` makes any alignment guarantees. But an `in` contract should be used here, if this solution is chosen.

> 2. You can check data.ptr and select the codepath for aligned or unaligned data at runtime. This will probably still have some performance impact (depending on the user code / data size passed to put). But you could combine this approach with version statements.

AFAIK even on x86, unaligned loads come with a performance penalty, although it's likely to be smaller for consecutive unaligned loads because of caching. Maybe provide a default implementation that can handle unaligned data, and an optimized one that requires aligned input?

jpf91 commented:
> I don't think @safe makes any alignment guarantees. But an in contract should be used here, if this solution is chosen.

Sorry, I should have added a more detailed explanation: Safe doesn't make any guarantees about alignment, but it guarantees memory safety. Unaligned loads can cause data and memory corruption on some architectures, therefore a function which might produce unaligned loads depending on input parameters could never be safe (or trusted). As this `put` function is not marked safe, it could be argued the API user is responsible for proper alignment (similar to the user being responsible for passing only zero terminated strings to C `str` functions). I don't think it's a good solution, but I think it's something that could be argued for. But you're right, an in contract is really necessary when using this approach.

> AFAIK even on x86, unaligned loads come with a performance penalty, although it's likely to be smaller for consecutive unaligned loads because of caching.

That's correct, but articles on that always compare `unaligned loads` vs. `aligned loads (of  pre-aligned data)`. But in this case we can't require pre-aligned data and the more interesting question is `unaligned loads` vs. `first copying data to an aligned buffer + aligned loads`. I haven't found any info on the performance of that. (In fact, a clever compiler might see that your just copying bytes to an aligned buffer and performing a load. So if the compiler concludes an unaligned load is faster it might actually rewrite the code.)

And as X86 uses the same instructions for unaligned vs. aligned loads IIRC there's shouldn't be any performance loss on X86 for aligned data when using the same codepath as for unaligned data. But this is just guessing. It's probably something that really needs benchmarks ;-)

> Maybe provide a default implementation that can handle unaligned data, and an optimized one that requires aligned input?

That's basically my solution 3. I think it's the best solution but I don't know how well it works with range composition (e.g. is `file.byChunk(1024)` an aligned buffer? It probably needs to go through a `map(a => cast(uint[])a` stage, etc)).

gchatelet commented:
Some numbers FYI based on code from [this repo](https://github.com/gchatelet/murmurhash3_d).

It's comparing the original C++ implementation with the _optimized_ D version and finally the digest version with the copy for alignment. I tested with the 3 compilers (help me tweak compilation flags if necessary).

Looking at the fastest implementation MurmurHash3_x64_128:
- For the optimized version: LDC runs at 93% of C++, GDC at 65%, DMD at 30%.
- For the aligned version: LDC runs at 8% of C++, GDC at 13%, DMD at 13%.

------

- DMD `DMD64 D Compiler v2.069.2`
```
 % make clean && make
rm -Rf build *.o
mkdir -p build
g++ -I/smasher/src -O3 -Wall -Werror -c smasher/src/MurmurHash3.cpp -o build/CMurmurHash3.o
dmd -O -inline -release benchmark.d murmurhash3.d build/CMurmurHash3.o -ofbuild/benchmark
./build/benchmark
Please wait while benchmarking MurmurHash3, running 4096*hash(256KiB) = 1GiB
C++ MurmurHash3_x64_128        - 19598 GiB/s
D SMurmurHash3_x64_128         - 5491 GiB/s
D digest MurmurHash3_x64_128   - 1561 GiB/s

C++ MurmurHash3_x86_128        - 12190 GiB/s
D SMurmurHash3_x86_128         - 3703 GiB/s
D digest MurmurHash3_x86_128   - 1352 GiB/s

C++ MurmurHash3_x86_32         - 9246 GiB/s
D SMurmurHash3_x86_32          - 3385 GiB/s
D digest MurmurHash3_x86_32    - 575 GiB/s
```

------

- GDC `gcc version 5.2.0 (crosstool-NG crosstool-ng-1.20.0-232-gc746732 - 20150830-2.066.1-dadb5a3784)`
```
 % make clean && make
rm -Rf build *.o
mkdir -p build
g++ -I/smasher/src -O3 -Wall -Werror -c smasher/src/MurmurHash3.cpp -o build/CMurmurHash3.o
~/Downloads/gdc_x86_64-pc-linux-gnu/bin/gdc -O3 -frelease -Wall -Werror benchmark.d murmurhash3.d build/CMurmurHash3.o -obuild/benchmark
./build/benchmark
Please wait while benchmarking MurmurHash3, running 4096*hash(256KiB) = 1GiB
C++ MurmurHash3_x64_128        - 19787 GiB/s
D SMurmurHash3_x64_128         - 12840 GiB/s
D digest MurmurHash3_x64_128   - 2586 GiB/s

C++ MurmurHash3_x86_128        - 12047 GiB/s
D SMurmurHash3_x86_128         - 8000 GiB/s
D digest MurmurHash3_x86_128   - 2279 GiB/s

C++ MurmurHash3_x86_32         - 9288 GiB/s
D SMurmurHash3_x86_32          - 7394 GiB/s
D digest MurmurHash3_x86_32    - 705 GiB/s
```

------

- LDC `LDC - the LLVM D compiler (0.16.1): based on DMD v2.067.1 and LLVM 3.7.0`
```
 % make clean && make
rm -Rf build *.o
mkdir -p build
g++ -I/smasher/src -O3 -Wall -Werror -c smasher/src/MurmurHash3.cpp -o build/CMurmurHash3.o
~/Downloads/ldc2-0.16.1-linux-x86_64/bin/ldc2 -O -release -inline benchmark.d murmurhash3.d build/CMurmurHash3.o -ofbuild/benchmark
./build/benchmark
Please wait while benchmarking MurmurHash3, running 4096*hash(256KiB) = 1GiB
C++ MurmurHash3_x64_128        - 18963 GiB/s
D SMurmurHash3_x64_128         - 17732 GiB/s
D digest MurmurHash3_x64_128   - 2609 GiB/s

C++ MurmurHash3_x86_128        - 12190 GiB/s
D SMurmurHash3_x86_128         - 10114 GiB/s
D digest MurmurHash3_x86_128   - 2346 GiB/s

C++ MurmurHash3_x86_32         - 9288 GiB/s
D SMurmurHash3_x86_32          - 8847 GiB/s
D digest MurmurHash3_x86_32    - 751 GiB/s
```

gchatelet commented:
> From your numbers, this should be 14%, right?

Good catch @klickverbot , I mismatch between LDC and DMD : )
For the aligned version: LDC runs at 13% of C++, GDC at 13%, DMD at 8%.

9il commented:
@gchatelet Can _manual_ inlining of `update` and `shufle` improve performance? There is no reason why GDC is slower than G++, but inlining may be bad.

jpf91 commented:
Use `-fno-bounds-check` and gdc is infinitely fast ;-) It seems without bounds checks all side effects in the benchmark gone and gdc removes the `hasher.putBlocks` and similar calls (That's what you get for marking everything pure ;-) Actually a quite nice result). So add something like this:

```d
//dummy.d
extern(C) void foo(ubyte[])
{
}
```

```d
//benchmark.d
extern(C) void foo(ubyte[]);

auto useHasher(H)() {
	H hasher;
	hasher.putBlocks(cast(const(H.Block)[])buffer);
	hasher.finalize();
	foo(hasher.getBytes()[]); //<=========================
	return hasher.getBytes();
}
```

```Makefile
$(BUILD_DIR)/benchmark: benchmark.d murmurhash3.d $(BUILD_DIR)/CMurmurHash3.o
	gdc dummy.d -c -o dummy.o
	gdc -O3 -frelease -Wall -Werror -fno-bounds-check dummy.o $^ -o$@
```

On my system results seem to be quite random for some reason, but `-fno-bounds-check` seems to improve performance for `SMurmurHash3_*`. So you might want to get rid of these boundchecks. 


I'm disappointed the compiler can't optimize this though:

```c
// gdc murmurhash3.d -fdump-tree-original -c
// cat murmurhash3.d.003t.original
;; Function putBlocks (_D11murmurhash320SMurmurHash3_x64_1289putBlocksMFNaNbNiNfMAxG2mXv)
;; enabled by -tree-original

{
  if (this != 0)
    {
      <<< Unknown tree: void_cst >>>
    }
  else
    {
      _d_assert_msg ({.length=9, .ptr="null this"}, {.length=13, .ptr="murmurhash3.d"}, 59);
    }
  {
    ulong block[2];
    ulong __key1526;
    struct const(ulong[2])[] __aggr1525;

    __aggr1525 = {.length=blocks.length, .ptr=blocks.ptr};
    __key1526 = 0;
    while (1)
      {
        if (!(__key1526 < __aggr1525.length)) break;
        block = *(__aggr1525.ptr + (__key1526 < __aggr1525.length ? __key1526 * 16 : _d_arraybounds ({.length=13, .ptr="murmurhash3.d"}, 61)));
        update ((ulong &) &this->h1, *(const ulong *) &block, this->h2, 9782798678568883157, 5545529020109919103, 31, 27, 1390208809);
        update ((ulong &) &this->h2, *((const ulong *) &block + 8), this->h1, 5545529020109919103, 9782798678568883157, 33, 31, 944331445);
        __key1526 = __key1526 + 1;
      }
  }
  this->size = this->size + blocks.length * 16;
}
```

jpf91 commented:
For the digest version it's definitely `Piecewise.put` which needs to be optimized. Changing `put` to only do `hasher.putBlocks(cast(const(Block)[])data);` for testing:
```
C++ MurmurHash3_x64_128        - 20898 GiB/s
D SMurmurHash3_x64_128         - 21904 GiB/s
D digest MurmurHash3_x64_128   - 21223 GiB/s

C++ MurmurHash3_x86_128        - 14949 GiB/s
D SMurmurHash3_x86_128         - 13128 GiB/s
D digest MurmurHash3_x86_128   - 15004 GiB/s

C++ MurmurHash3_x86_32         - 9225 GiB/s
D SMurmurHash3_x86_32          - 9288 GiB/s
D digest MurmurHash3_x86_32    - 9330 GiB/s
```
(As already said test results on my machine seem to be kinda weird...)

schuetzm commented:
> Unaligned loads can cause data and memory corruption on some architectures

I wasn't aware of that, I thought it would always trap.

> > Maybe provide a default implementation that can handle unaligned data, and an optimized one that requires aligned input?
> 
> That's basically my solution 3. I think it's the best solution but I don't know how well it works with range composition (e.g. is file.byChunk(1024) an aligned buffer? It probably needs to go through a map(a => cast(uint[])a stage, etc)).

I was thinking of `MurmurHash3` vs `MurmurHash3Aligned`, both with the usual `void put(scope const(ubyte)[])` methods.

gchatelet commented:
Thx @jpf91 , I updated the benchmark but kept the unaligned version.

GDC even [beats the C++ implementation](http://e.lvme.me/l2fdwg1.jpg) (more inlining possibilities I guess)


```
 % make clean && make -j12
rm -Rf build *.o
mkdir -p build
g++ -I/smasher/src -O3 -Wall -Werror -c smasher/src/MurmurHash3.cpp -o build/CMurmurHash3.o
~/Downloads/gdc_x86_64-pc-linux-gnu/bin/gdc -O3 -frelease -fno-bounds-check -Wall -Werror benchmark.d murmurhash3.d build/CMurmurHash3.o -obuild/gdc_benchmark
~/Downloads/ldc2-0.16.1-linux-x86_64/bin/ldc2 -O5 -release -inline benchmark.d murmurhash3.d build/CMurmurHash3.o -ofbuild/ldc_benchmark
dmd -O -inline -release benchmark.d murmurhash3.d build/CMurmurHash3.o -ofbuild/dmd_benchmark
build/gdc_benchmark
Please wait while benchmarking MurmurHash3, running 4096*hash(256KiB) = 1GiB
C++ MurmurHash3_x64_128        -  94% - 18789 GiB/s
D SMurmurHash3_x64_128         - 100% - 19787 GiB/s
D digest MurmurHash3_x64_128   -  12% - 2535 GiB/s
C++ MurmurHash3_x86_128        -  61% - 12083 GiB/s
D SMurmurHash3_x86_128         -  63% - 12488 GiB/s
D digest MurmurHash3_x86_128   -  12% - 2377 GiB/s
C++ MurmurHash3_x86_32         -  46% - 9288 GiB/s
D SMurmurHash3_x86_32          -  45% - 8943 GiB/s
D digest MurmurHash3_x86_32    -   3% - 710 GiB/s
build/ldc_benchmark
Please wait while benchmarking MurmurHash3, running 4096*hash(256KiB) = 1GiB
C++ MurmurHash3_x64_128        - 100% - 19692 GiB/s
D SMurmurHash3_x64_128         -  89% - 17579 GiB/s
D digest MurmurHash3_x64_128   -  12% - 2368 GiB/s
C++ MurmurHash3_x86_128        -  60% - 12012 GiB/s
D SMurmurHash3_x86_128         -  51% - 10114 GiB/s
D digest MurmurHash3_x86_128   -  11% - 2173 GiB/s
C++ MurmurHash3_x86_32         -  47% - 9288 GiB/s
D SMurmurHash3_x86_32          -  44% - 8809 GiB/s
D digest MurmurHash3_x86_32    -   3% - 687 GiB/s
build/dmd_benchmark
Please wait while benchmarking MurmurHash3, running 4096*hash(256KiB) = 1GiB
C++ MurmurHash3_x64_128        - 100% - 19598 GiB/s
D SMurmurHash3_x64_128         -  27% - 5469 GiB/s
D digest MurmurHash3_x64_128   -   7% - 1560 GiB/s
C++ MurmurHash3_x86_128        -  62% - 12154 GiB/s
D SMurmurHash3_x86_128         -  18% - 3703 GiB/s
D digest MurmurHash3_x86_128   -   6% - 1371 GiB/s
C++ MurmurHash3_x86_32         -  47% - 9267 GiB/s
D SMurmurHash3_x86_32          -  17% - 3396 GiB/s
D digest MurmurHash3_x86_32    -   3% - 597 GiB/s
```

jpf91 commented:
@schuetzm 
>>Unaligned loads can cause data and memory corruption on some architectures

> I wasn't aware of that, I thought it would always trap.

I had to debug such a problem for GDC once ;-) See for example: http://infocenter.arm.com/help/index.jsp?topic=/com.arm.doc.dui0068b/Bcfihdhj.html
> __Address alignment for word transfers__
> In most circumstances, you must ensure that addresses for 32-bit transfers are 32-bit word-aligned.
>[...]
> If your system does not have a system coprocessor (cp15), or alignment checking is disabled:
> * For STR, the specified address is rounded down to a multiple of four. [effectively overwriting data at a lower address => data corruption]

The same behavior always happens on old ARMv5 chips. For loads it's more complicated. A load will not corrupt memory, but will load invalid data. If this data is then somehow used for memory address calculation you can still end up with corrupted memory.
If you want to read more about this, see http://www.heyrick.co.uk/armwiki/Unaligned_data_access and http://stackoverflow.com/a/16549366/471401 .

> I was thinking of MurmurHash3 vs MurmurHash3Aligned, both with the usual void put(scope const(ubyte)[]) methods.

That might be a good idea. Some testing shows that on `X86_64` alignment is less of a problem than data length though, and `ubyte[]` allows any data length. I think `MurmurHash3Aligned` with `put(Block[])` is more useful. But then there's no big difference to `SMurmurHash3_*`.


@gchatelet 

* LDC and DMD would also probably perform better when using the noboundscheck option. However, for phobos the code will have to be compiled with boundschecks, so you'll have to change the code to workaround bounds checking code:
 ```d
 void putBlocks(scope const(Block)[] blocks...) pure nothrow @nogc @trusted
 {
     const(Block)* start = blocks.ptr;
     const(Block)* end = blocks.ptr + blocks.length; 

     for(auto ptr = start; ptr < end; ptr++)
     {
         update(h1, (*ptr)[0], h2, c1, c2, 31, 27, 0x52dce729);
         update(h2, (*ptr)[1], h1, c2, c1, 33, 31, 0x38495ab5);
     }
     size += blocks.length * Block.sizeof;
 }
```
* Aligned vs. unaligned data seems to perform more or less the same for me on `x86_64` when using `putBlocks`
* With GDC there's no slowdown on `X86_64` when first reading into a buffer. This code is safe for architectures without unaligned loads. For DMD there's a 3% slowdown.
 ```d
 struct Piecewise(Hasher)
 {
     union BufferUnion
     {
         Block block;
         ubyte[Block.sizeof] blockData;
     }
     BufferUnion buffer;

     void put(scope const(ubyte)[] data...) pure nothrow @trusted
     {
         // Buffer should never be full while entring this function.
         assert(bufferSize < Block.sizeof);
         const(ubyte)* start = data.ptr;

         size_t preprocessed = 0;
         /*
          * Do main work: process chunks of Block.sizeof bytes
          */
         size_t numBlocks = (data.length - preprocessed) / Block.sizeof;
         const(ubyte)* end = start + numBlocks * Block.sizeof;
         
         for(; start < end; start += Block.sizeof)
         {
             buffer.blockData = start[0 .. Block.sizeof];
             hasher.putBlock(buffer.block);
         }
         hasher.size += numBlocks * Block.sizeof;
     }
 }

 struct SMurmurHash3_x64_128
 {
     void putBlock(scope const Block block) pure nothrow @nogc
     {
         update(h1, block[0], h2, c1, c2, 31, 27, 0x52dce729);
         update(h2, block[1], h1, c2, c1, 33, 31, 0x38495ab5);
     }
 }
 ```
* There is a 10-20% slowdown when supporting data buffers with lengths which are not multiples of the block size. Here's my code: Note: This code is untested and needs proof-reading as well as many tests, especially for corner cases. (`data.length < Block.sizeof`, `data.length == Block.sizeof`, ...). The `if (data.length + bufferSize < Block.sizeof)` statement seems to be the main slowdown source.
 ```d
 struct Piecewise(Hasher)
 {
     union BufferUnion
     {
         Block block;
         ubyte[Block.sizeof] blockData;
     }
     BufferUnion buffer;

     void put(scope const(ubyte)[] data...) pure nothrow @trusted
     {
         // Buffer should never be full while entring this function.
         assert(bufferSize < Block.sizeof);
         const(ubyte)* start = data.ptr;


         /*
          * Check if we have some leftover data in the buffer.
          * Then fill the first block buffer.
          */
         if (data.length + bufferSize < Block.sizeof)
         {
             buffer.blockData[bufferSize .. $] = start[0 .. data.length];
             bufferSize += data.length;
             return;
         }
         auto preprocessed = Block.sizeof - bufferSize;
         buffer.blockData[bufferSize .. $] = start[0 .. preprocessed];
         hasher.putBlock(buffer.block);
         start += preprocessed;

         /*
          * Do main work: process chunks of Block.sizeof bytes
          */
         size_t numBlocks = (data.length - preprocessed) / Block.sizeof;
         const(ubyte)* end = start + numBlocks * Block.sizeof;
         
         for(; start < end; start += Block.sizeof)
         {
             buffer.blockData = start[0 .. Block.sizeof];
             hasher.putBlock(buffer.block);
         }
         // +1 for preprocessed Block
         hasher.size += (numBlocks + 1) * Block.sizeof;

         /**
          * Now add remaining data to buffer
          */
         bufferSize = data.length - preprocessed - numBlocks * Block.sizeof;
         buffer.blockData[0 .. bufferSize] = end[0 .. bufferSize];
     }
 }

 struct SMurmurHash3_x64_128
 {
     void putBlock(scope const Block block) pure nothrow @nogc
     {
         update(h1, block[0], h2, c1, c2, 31, 27, 0x52dce729);
         update(h2, block[1], h1, c2, c1, 33, 31, 0x38495ab5);
     }
 }
 ```
* At least for `X86_64` we can ignore the effects of unaligned data and the overhead caused by copying. So the copying version which will be safe for all architectures has no negative effects on `X86_64`. A much bigger problem is processing data buffers with lengths which are not multiples of the Blocksize. So maybe a `MurmurHash3Aligned` digest which only accepts `put(ulong[Block.sizeof/8][])` is a good idea for those use cases where we have proper data sizes. (I'm not actually sure what type we should use for that. `ubyte[n]` would be nice, but `ubyte[n].alignof == 1` so that's not what we really want)

* Here are ARM results for ARMv5 (no unaligned load support) with linux kernel configured to fix up these accesses (very slow):
 ```
 C++ MurmurHash3_x64_128, aligned data -  97% - 529 GiB/s
 C++ MurmurHash3_x64_128, unaligned data -   6% - 33 GiB/s
 D SMurmurHash3_x64_128, aligned data - 100% - 544 GiB/s
 D SMurmurHash3_x64_128, unaligned data -   5% - 33 GiB/s
 D digest MurmurHash3_x64_128, aligned data -  57% - 311 GiB/s
 D digest MurmurHash3_x64_128, unaligned data -  53% - 293 GiB/s
 ```
Please note that the `C++ MurmurHash3_x64_128` and `SMurmurHash3_x64_128` generate invalid results for unaligned data with the standard linux kernel configuration. The digest version works fine. So an overload for a fast version for aligned data is even more important for ARM. I couldn't find a specific reason for the 40% slowdown for the digest version. Even simply forwarding the data by calling `putBlocks` is 10% slower. GCC probably can't optimize ARM code as well as X86_64 code.


My conclusion would be to use the unaligned-data-safe but slightly slower approach I've posted above for the default `MurmurHash3_x86_32` `std.digest` digests. Then either introduce `MurmurHash3_x86_32Aligned` with a `put` function which only takes a `Block[]` parameter. Or even better: maybe make the `SMurmurHash3_*` structs compatible with the `digest` API (alias `put` => `putBlocks`, add `start` and `finish` functions).

gchatelet commented:
Here are the numbers with the latest version. I used @jpf91 's suggestions and added more tests. `dmd` is still super slow but `ldc` and `gdc` are now as fast as the C++ version with bounds-checking. Also `ldc` is now benchmarked against C++ code compiled with `clang++` to be fair. Last thing I fixed the benchmark's results to display correct throughput (should have been MiB instead of GiB).

I'm now working on improving the documentation, stay tuned !
Feedback is still welcome of course :)

```
 % make clean && make -j12
rm -Rf build *.o
mkdir -p build
clang++ -I/smasher/src -O3 -Wall -Werror -c smasher/src/MurmurHash3.cpp -o build/CMurmurHash3_clang.o
g++ -I/smasher/src -O3 -Wall -Werror -c smasher/src/MurmurHash3.cpp -o build/CMurmurHash3_g++.o
dmd -O -release -inline benchmark.d murmurhash3.d build/CMurmurHash3_g++.o -ofbuild/dmd_benchmark
~/Downloads/gdc_x86_64-pc-linux-gnu/bin/gdc -O3 -frelease -Wall -Werror -march=native benchmark.d murmurhash3.d build/CMurmurHash3_g++.o -obuild/gdc_benchmark
~/Downloads/ldc2-0.16.1-linux-x86_64/bin/ldc2 -O3 -release -inline benchmark.d murmurhash3.d build/CMurmurHash3_clang.o -ofbuild/ldc_benchmark
build/gdc_benchmark
Please wait while benchmarking MurmurHash3, running 4096*hash(256KiB) = 1GiB
x64_128 C++          - 100% - 19.51 GiB/s
x64_128 D            - 105% - 20.49 GiB/s
x64_128 D digest     -  89% - 17.53 GiB/s

x86_128 C++          - 100% - 12.02 GiB/s
x86_128 D            -  93% - 11.31 GiB/s
x86_128 D digest     -  89% - 10.75 GiB/s

x86_32 C++           - 100% - 9.26 GiB/s
x86_32 D             -  97% - 9.04 GiB/s
x86_32 D digest      - 102% - 9.46 GiB/s

build/ldc_benchmark
Please wait while benchmarking MurmurHash3, running 4096*hash(256KiB) = 1GiB
x64_128 C++          - 100% - 15.57 GiB/s
x64_128 D            - 113% - 17.64 GiB/s
x64_128 D digest     -  92% - 14.32 GiB/s

x86_128 C++          - 100% - 9.14 GiB/s
x86_128 D            - 110% - 10.11 GiB/s
x86_128 D digest     -  97% - 8.89 GiB/s

x86_32 C++           - 100% - 8.79 GiB/s
x86_32 D             -  99% - 8.74 GiB/s
x86_32 D digest      - 104% - 9.20 GiB/s

build/dmd_benchmark
Please wait while benchmarking MurmurHash3, running 4096*hash(256KiB) = 1GiB
x64_128 C++          - 100% - 19.74 GiB/s
x64_128 D            -  23% - 4.59 GiB/s
x64_128 D digest     -  23% - 4.66 GiB/s

x86_128 C++          - 100% - 12.15 GiB/s
x86_128 D            -  25% - 3.04 GiB/s
x86_128 D digest     -  25% - 3.06 GiB/s

x86_32 C++           - 100% - 9.28 GiB/s
x86_32 D             -  36% - 3.37 GiB/s
x86_32 D digest      -  36% - 3.36 GiB/s
```

jpf91 commented:
@gchatelet That looks great! Thanks for incorporating the suggestions and for contributing this module.

Some small comments:
* Maybe add a comment like this to `putBlocks` and `Piecewise.put`:

 ```d
 /*
  * Use pointer manipulation instead of foreach to avoid bounds checking (see @@@BUG@@@ 15581).
  * This function heavily affects the performance of [Piecewise/hash calculation] so make sure to
  * benchmark all changes!
  */
 ```
* Maybe rename the module to `murmurhash` instead of `murmurhash3`. All other modules do not include the algorithm number. (If we add murmurhash2 at some point we could still put these into extra files and use `package.d` for `std.digest.murmurhash`).
* You need to have another look at the windows Makefiles and add `murmurhash.d` to some more rules. Just search for `crc.d` and add the same rules.


I did a quick review and apart from these small issues and the missing documentation this module definitely gets a :+1: from me. The code looks quite nice. The `checkResult` unittest helper could be useful for other `std.digest` modules. Maybe also use the dmd `-cov` coverage check and remember to squash your commits when this pull request is ready for merging.

gchatelet commented:
Can someone have a look on the [last version of the documentation](http://dtest.thecybershadow.net/artifact/website-a5f4d2335772b8c008372fc71b4983a3787afe25-31cd26ef413e30f4fcd8d659e2afae3b/web/library-prerelease/std/digest/murmurhash.html) and give some feedback?

Also I've been told to squash the history to produce a single commit. Since a rebase with squashing rewrites the history I needed to `push` with `--force` which seems to confuse the auto-tester (search for `1d60afc` on this page to pin down the rewritten commit). Was it the expected way of doing it? Or shall I create a new pull request when everything is good?

Finally I'd need a permission to merge if everything looks good.

JakobOvrum commented on std/digest/murmurhash.d:
```diff
@@ -0,0 +1,789 @@
+@safe:
+
+///
+unittest
```
DDox recognizes these two examples, but DDoc doesn't seem to. Maybe they need to be moved up?

9il replied:
Please move this unittest right after the module declaration.

yebblies commented:
> seems to confuse the auto-tester

We don't worry about the results on old commits.  As long as https://auto-tester.puremagic.com/pull-history.ghtml?projectid=1&repoid=3&pullid=3916 is showing the right latest commit.

If you fetch and rebase on top of upstream/master (or edit the first commit) those commit hashes will all change and the old results will disappear anyway.

> Was it the expected way of doing it?

Yes.

> Or shall I create a new pull request when everything is good?

No, rebasing is correct.

jpf91 commented:
I think you can move the
```
Implementation uses pointer manipulation instead of foreach to avoid bounds checking (see @@@BUG@@@ 15581) making the function @trusted instead of @safe. This function heavily affects the performance of hash calculation so make sure to benchmark all changes. Benchmark can be found on github.
```
 documentation into a normal comment, it doesn't really need to appear in DDOC.

And the `putBlock`, `putRemainder`, `finalize` and `get` methods of the `SMurmurHash3_*` structs should be documented (a single summary sentence per function). Otherwise LGTM :+1:

JakobOvrum commented on std/digest/murmurhash.d:
```diff
@@ -0,0 +1,789 @@
+to benchmark all changes. Benchmark can be found on
+$(WEB https://github.com/gchatelet/murmurhash3d, github).
+*/
+void putBlocks(H, Block = H.Block)(ref H hasher, scope const(Block[]) blocks...) pure nothrow @nogc @trusted
```
This is misuse of `@trusted` - don't put it on templated functions which memory safety depends on template arguments. This silently breaks `@safe`'s promise of memory safety when `H.putBlock` can incur memory errors. I've written more about this [here](https://jakobovrum.github.io/d/2016/01/20/memory-safety.html).

Is `H` assumed to be a struct from this module? If so, and if all their `putBlock` functions have been manually vetted for memory safety when called in this manner, then I think it's fine as long as a template constraint is added to enforce the assumption. That way, all possible instantiations of `putBlocks` are memory safe, so it can correctly be annotated with `@trusted`.

However, if the `putBlock` functions all have a memory safe API, the approach using nested `@trusted` functions described in the aforementioned article is probably better, as it leaves `@trusted` to each individual `putBlock` implementation, making memory safety easier to review. Of course we'd have to see how it works with the benchmarks.

Also, if you don't want to deal with using `@trusted` correctly right now, feel free to leave things `@system`. It's better to leave memory safety guarantees to future work than violating `@safe`.

gchatelet replied:
Thx for pointing this out.
I added a `@trusted` nested function. Benchmark showed no change in performance.

JakobOvrum commented on std/digest/murmurhash.d:
```diff
@@ -0,0 +1,789 @@
+        {
+            putBlock(*start);
+        }
+        size += blocks.length * Block.sizeof;
```
I'm not that familiar with the digest package, but what is the `size` property for? It doesn't appear to be read in this module, and I don't see any mention of it in the package documentation.

gchatelet replied:
`size` is used during finalization step to get the array length in the mix.
You can look  for `len` on the [wikipedia page](https://en.wikipedia.org/wiki/MurmurHash).

JakobOvrum commented on std/digest/murmurhash.d:
```diff
@@ -0,0 +1,789 @@
+    to benchmark all changes. Benchmark can be found on
+    $(WEB https://github.com/gchatelet/murmurhash3d, github).
+    */
+    void put(scope const(ubyte[]) data...) pure nothrow @trusted
```
This has the same issue with `@trusted` (arbitrary code is injected through `this.hasher`)

gchatelet replied:
Unfortunately the nested function trick destroys the performances here. Maybe because `put` is a member function. I have to make the function `@system` for now. I'll leave enough comment for later cleanup.

gchatelet commented:
Thx everyone for the advices. I think this is good to go.

gchatelet commented:
Ping :)

JakobOvrum commented on std/digest/murmurhash.d:
```diff
@@ -0,0 +1,815 @@
+    // sure to benchmark all changes. Benchmark can be found at
+    // https://github.com/gchatelet/murmurhash3d.
+    // TODO(gchatelet) revert to @safe and foreach when @@@BUG@@@ 15581 is fixed.
+    scope auto putBlockTrusted = (ref H hasher, scope const(Block[]) blocks...) pure nothrow @nogc @trusted
```
This has exactly the same problem as before as `putBlock` and `size` (could be a `@system` property function) are in `@trusted`.

gchatelet replied:
hmm so I didn't understand what you meant in the first place then. Let me read your article first.

DmitryOlshansky commented:
Looks great! Ping @9il @jpf91 @schuetzm

gchatelet commented:
Since D-Programming-Language/dmd#5470 is in I intend to simplify the code to have everything @safe.
I started but got side tracked I'll push a new version later today.

schuetzm commented:
Some questions:
* Why are the structs called `SMurmurHash_...` instead of just `MurmurHash_...`?
* There should be an alias called `(S)MurmurHash` that is defined as the appropriate implementation for the current architecture.
* If `Piecewise` is "not intended to be used directly" as the documentation says, it should be private.

WalterBright commented on std/digest/murmurhash.d:
```diff
@@ -0,0 +1,815 @@
+Note: The current implementation is optimized for little endian architectures.
+It will exhibit different results on big endian architectures and a slightly less uniform distribution.
+
+License: $(WEB www.boost.org/LICENSE_1_0.txt, Boost License 1.0).
```
Missing copyright

gchatelet replied:
@WalterBright is this ok? (I borrowed it from another file)
BTW it looks like not all files have copyright.

gchatelet commented:
Erf sorry I completely mess up the pull request :(
Let me try to fix this.

gchatelet commented:
@schuetzm 

> Why are the structs called SMurmurHash_... instead of just MurmurHash_...?

`SMurmurHash_` is the low level struct used by people knowing what they're doing.
`MurmurHash_` is an alias to `Piecewise!SMurmurHash_` it allows to push data that is not a multiple of block size and is suitable for use by the `digest` library.
I'm happy to change the naming to something better, suggestions welcome.

> There should be an alias called (S)MurmurHash that is defined as the appropriate implementation for the current architecture.

I'm not sure to understand what you mean. Please see my comment above.

> If Piecewise is "not intended to be used directly" as the documentation says, it should be private.

It is private in the sense that it can be instantiated only with types in the same module but it has to be public to be used by the digest library.

schuetzm commented:
I was going by the documentation: http://dtest.thecybershadow.net/artifact/website-c30f31be957f108026d0684c62a1cda24b5d949f-fa589e0c6067cac4be3e0bffc3f9769b/web/phobos-prerelease/std_digest_murmurhash.html

It lists many architecture specific `MurmurHash3_x86/x64_xyz` variants, but it doesn't mention a generic version that can be used on all architectures. I don't think anything architecture specific should be exposed at all, what would be the use? It's only an implementation detail. Instead, there should be `MurmurHash3_32` and `MurmurHash3_128` aliases pointing to the right implementation.

As for `Piecewise`, are you sure it has to be public? In other words, did you try to make it private? If that doesn't work, I'd consider that a compiler bug. In any case, if it isn't supposed to be used outside the module, there should be no (ddoc) documentation.

andralex commented:
Soooo...

* Why `S` as a prefix to structs? We don't do that.
* As a putative user, how do I compute murmur hash portably? All I see is names with `_x86` in them so I worry it's not portable to other architectures.

I'd just expect some `struct MurmurHash(uint size)` with `size` restricted to 32, 64, or 128.

9il commented:
> As a putative user, how do I compute murmur hash portably? All I see is names with _x86 in them so I worry it's not portable to other architectures.

Each variant would work on any architecture. This historical suffix, which refers to optimization pattern. Maybe we can change name to `_opt32bit` and `_opt64bit`

gchatelet commented:
I'm abroad with very low connectivity until April 10.
-  Sure we can get rid of the S prefix. What do you suggest for naming to differentiate between the low level API and the digest API? I can't come up with a good name.
- I would stick with the original naming since each variant produce different results but are optimized for different platforms. The JavaScript implementation as the same names for instance. We can still provide an alias that select the best version according to the platform but then people might expect the same result on x32 and x86_64 which will not be the case. Thoughts ?

jpf91 commented:
> We can still provide an alias that select the best version according to the platform but then people might expect the same result on x32 and x86_64 which will not be the case. Thoughts ?

I think if we provide such an alias, we have to make it very obvious that it produces different hashes on different systems. `UnportableMurmurHash3` or `LocalMurmurHash3` or something like that.

9il commented:
Rebase please

gchatelet commented:
@9il just rebased.
FYI I'll be at DConf2016, it might be easier to discuss the last points of this PR IRL.

gchatelet commented:
Please take another look.

I updated the documentation, changed the naming of the structure and improved the performance of DMD a bit by returning by value in ```update(T)(...)``` instead of passing by ref.

Let me know if there's anything else I can do to improve this PR.

aappleby commented on std/digest/murmurhash.d:
```diff
@@ -0,0 +1,850 @@
+unittest
+{
+    // dfmt off
+    checkResult!FastMurmurHash3_x86_32([
```
If you're going to add verification values, take a look at how SMHasher does things in VerificationTest() -

https://github.com/aappleby/smhasher/blob/master/src/KeysetTest.cpp
https://github.com/aappleby/smhasher/blob/master/src/main.cpp

9il commented on std/digest/murmurhash.d:
```diff
@@ -0,0 +1,850 @@
+    @disable this(this);
+
+    /// Adds a single Block of data without increasing size.
+    /// Make sure to increase size by Block.sizeof for each call to putBlock.
```
Please use `/++ +/` for multiline comments

9il commented:
LGTM! Thank you!

9il commented:
please add changelog info and update index

schuetzm commented:
About that x86/x64 distinction: The [Wikipedia page](https://en.wikipedia.org/wiki/MurmurHash#Variants) calls these variants `MurmurHash64A` and `MurmurHash64B`. I would much prefer these names. There's really nothing x86(_64) specific about them.

Also, the documentation says "MurmurHash3 optimized for x86 processors". Is this true? Or is it really optimized for 32bit architectures in general, rather than x86 specifically? The source code looks completely generic, so I suspect it's the latter. Please adjust the documentation accordingly.

9il commented:
> Also, the documentation says "MurmurHash3 optimized for x86 processors". Is this true? Or is it really optimized for 32bit architectures in general, rather than x86 specifically? The source code looks completely generic, so I suspect it's the latter. Please adjust the documentation accordingly.
>  Hide all checks

Yes, implementation is generic. MurmurHash3 was optimized for x86 (A for 32 bit and B for 64 bit) as an algorithm.

gchatelet commented:
@9il can you point me to relevant documentation for `changelog info and update index` ?
@schuetzm , as @aappleby mentioned I think we should have a generic test suite for compliance and speed for all the hashes in `std.digest`. It would be nice to track performance for algorithm/compiler/architecture.

wilzbach commented:
> @9il can you point me to relevant documentation for changelog info and update index ?

https://github.com/dlang/phobos/blob/master/changelog.dd
It's a bit tricky, but you need to ensure that the RELATIVE_LINK2 name that you add is the same - it will soon get easier to add stuff to the changelog.
I couldn't find an index page for digest - @9il?

>  think we should have a generic test suite for compliance and speed for all the hashes in std.digest. It would be nice to track performance for algorithm/compiler/architecture.

Or better: for entire Phobos ;-)

gchatelet commented:
> Or better: for entire Phobos ;-)

@wilzbach totally agree. We should provide a way to spot phobos performance regressions. But that's for a separate PR :-)

9il commented:
> @9il can you point me to relevant documentation for changelog info and update index ?
> https://github.com/dlang/phobos/blob/master/changelog.dd
> It's a bit tricky, but you need to ensure that the RELATIVE_LINK2 name that you add is the same - it will soon get easier to add stuff to the changelog.
> I couldn't find an index page for digest - @9il?

https://github.com/dlang/phobos/blob/master/index.d

gchatelet commented:
Almost there. Before moving forward I would like to make sure the naming is correct.
Marc pointed out the x86/x64 part was confusing. I added some documentation about that but still.

The `A` and `B` variants are only for MurmurHash2 IIUC.
How about the following?
- `MurmurHash3_32`
- `MurmurHash3_128`
- `MurmurHash3_128_opt_x64`

Suggestions are welcome.

I think the test suite / benchmark from SMHasher should be integrated part of `std.digest` so it should be a different PR.

9il commented:
Yes, x86 must go.

schuetzm commented:
Well, almost. It still says `x64`, but isn't specific to `x64`. There are tons of other 64-bit architectures. And for symmetry reasons, maybe rename the 32-bit optimized version to `MurmurHash3_128_opt_32`, too?

9il commented:
> MurmurHash3_32
> MurmurHash3_128
> MurmurHash3_128_opt_x64

`MurmurHash3_32_opt32`
`MurmurHash3_128_opt32`
`MurmurHash3_128_opt64`

gchatelet commented:
@andralex can you take a final look?

wilzbach commented:
> @andralex can you take a final look?

You can't reach Andrei via Github (he get's way too many notifications), write him a short mail to ping him ;-)

andralex commented:
Thanks for the email. Will get on this soon.

andralex commented on posix.mak:
```diff
@@ -170,7 +170,7 @@ PACKAGE_std_experimental = typecons
   sorting
 PACKAGE_std_container = array binaryheap dlist package rbtree slist util
-PACKAGE_std_digest = crc digest hmac md ripemd sha
+PACKAGE_std_digest = crc digest hmac md ripemd sha murmurhash
```
These should ome in alphabetic order.

andralex commented on std/digest/murmurhash.d:
```diff
@@ -0,0 +1,860 @@
+/**
+Computes MurmurHash hashes of arbitrary data. MurmurHash is a non-cryptographic
```
Link the first occurrence of the word MurmurHash to the appropriate Web reference.

andralex commented on std/digest/murmurhash.d:
```diff
@@ -0,0 +1,860 @@
+/**
+Computes MurmurHash hashes of arbitrary data. MurmurHash is a non-cryptographic
+hash function suitable for general hash-based lookup. It is optimized for x86
+but can be used on every architectures.
```
every architectures -> all architectures

andralex commented on std/digest/murmurhash.d:
```diff
@@ -0,0 +1,860 @@
+The current version is MurmurHash3, which yields a 32-bit or 128-bit hash value.
+The older MurmurHash 1 and 2 are currently not supported.
+
+MurmurHash3 comes in three flavors, providing greater and greater throughput:
```
providing greater and greater throughput -> listed in increasing order of throughput

andralex commented on std/digest/murmurhash.d:
```diff
@@ -0,0 +1,860 @@
+
+    /++
+    Adds a single Block of data without increasing size.
+    Make sure to increase size by Block.sizeof for each call to putBlock.
```
Use backquotes around all code fragments, such as Block.sizeof and putBlock.

andralex commented on std/digest/murmurhash.d:
```diff
@@ -0,0 +1,860 @@
+    @disable this(this);
+
+    /++
+    Adds a single Block of data without increasing size.
```
Why is Block capitalized?

andralex commented on std/digest/murmurhash.d:
```diff
@@ -0,0 +1,860 @@
+        h1 = update(h1, block, 0, c1, c2, 15, 13, 0xe6546b64U);
+    }
+
+    /// Put remainder bytes. This must be called only once after putBlock and before finalize.
```
backquotes

andralex commented on std/digest/murmurhash.d:
```diff
@@ -0,0 +1,860 @@
+
+    /++
+    Pushes an array of blocks at once. It is more efficient to push as much data as possible in a single call.
+    On platform that does not support unaligned reads (MIPS or old ARM chips), the compiler may produce slower code to ensure correctness.
```
platform -> platforms

andralex replied:
does not -> do not

andralex commented on std/digest/murmurhash.d:
```diff
@@ -0,0 +1,860 @@
+    }
+
+    /// Returns the current hashed value as an ubyte array.
+    auto getBytes() pure nothrow @nogc
```
Would be probably nice to specify the return type so it appears in clear to the reader.

andralex commented on std/digest/murmurhash.d:
```diff
@@ -0,0 +1,860 @@
+    /// Returns the current hashed value as an ubyte array.
+    auto getBytes() pure nothrow @nogc
+    {
+        return cast(ubyte[Block.sizeof]) cast(uint[1])[get()];
```
... in which case you can use `cast (typeof(return))` here

andralex commented on std/digest/murmurhash.d:
```diff
@@ -0,0 +1,860 @@
+{
+    import std.string : representation;
+
+    auto hash(H, Block = H.Block)(string data)
```
private

andralex commented on std/digest/murmurhash.d:
```diff
@@ -0,0 +1,860 @@
+        return hasher.getBytes();
+    }
+
+    void checkResult(H)(in string[string] groundtruth)
```
private

andralex commented on std/digest/murmurhash.d:
```diff
@@ -0,0 +1,860 @@
+remainder bytes in a buffer and pushes it when the block is complete or during
+finalization.
+*/
+struct Piecewise(Hasher) if (moduleName!Hasher == "std.digest.murmurhash")
```
this should be private then

andralex commented on std/digest/murmurhash.d:
```diff
@@ -0,0 +1,860 @@
+    enum blockSize = bits!Block;
+
+    alias Block = Hasher.Block;
+    union BufferUnion
```
Because of the `union` you'd need to add the occasional `@trusted` to the users of this structure.

andralex commented on std/digest/murmurhash.d:
```diff
@@ -0,0 +1,860 @@
+    // Initialize
+    void start()
+    {
+        this = Piecewise.init;
```
`this = this.init;` so it doesn't depend on the name

andralex commented on std/digest/murmurhash.d:
```diff
@@ -0,0 +1,860 @@
+        // Do main work: process chunks of Block.sizeof bytes.
+        const numBlocks = data.length / Block.sizeof;
+        const remainderStart = numBlocks * Block.sizeof;
+        foreach (const Block block; cast(const(Block[]))(data[0 .. remainderStart]))
```
`ref`?

andralex commented on std/digest/murmurhash.d:
```diff
@@ -0,0 +1,860 @@
+    assert(digester.getRemainder() == []);
+}
+
+private template bits(T)
```
this is not really deserving a name

andralex commented:
I added a few nits, and have a high-level question. So we have separate structures defined for "optimized for 32 bits systems" and "optimized for 64 bits systems". Why aren't those hidden and selected automatically depending on the platform? Would someone be interested in running the opt64 thing on a 32-bit system or vice versa?

andralex commented on std/digest/murmurhash.d:
```diff
@@ -0,0 +1,860 @@
+    uint h1;
+
+public:
+    alias Block = uint; /// The element type for 32-bit implementation.
```
I think the name `Block` is really confusing. I've wondered what that is all through the code review. Would `Unit` fare better?

gchatelet replied:
@andralex I reused the terminology from the [original implementation](https://github.com/aappleby/smhasher/blob/master/src/MurmurHash3.cpp). I can go for `Unit` if you prefer.

andralex commented:
Naming-wise, I think we should go with one only:

```d
struct MurmurHash(uint size /* 32, 64, or 128 */, uint opt = size_t.sizeof == 8 ? 64 : 32);
```

9il commented:
> Would someone be interested in running the opt64 thing on a 32-bit system or vice versa?

Yes, this is common for distributed databases.

andralex commented:
@9il thanks, cool. Then it seems the template with the defaulted second argument is appropriate.

gchatelet commented:
I posted a new version based on @andralex suggestions. It's more compact and clearer.
One drawback is that we can't distinguish between MurmurHash2 32 bit (unimplemented) and MurmurHash3 32 bit.
Maybe we just don't care...
Let's just hope there won't be a MurmurHash4 : )

Destroy!

schuetzm commented:
The name should still be `MurmurHash3`, because as you said `MurmurHash2` is a thing, too. The module can still be called `murmurhash` without a version number.

aappleby commented:
I have ideas for a MurmurHash4, but I don't know if I'll ever publish it. :)

On Fri, May 27, 2016 at 2:57 AM, Guillaume Chatelet <
notifications@github.com> wrote:

> I posted a new version based on @andralex <https://github.com/andralex>
> suggestions. It's more compact and clearer.
> One drawback is that we can't distinguish between MurmurHash2 32 bit
> (unimplemented) and MurmurHash3 32 bit.
> Maybe we just don't care...
> Let's just hope there won't be a MurmurHash4 : )
>
> Destroy!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/dlang/phobos/pull/3916#issuecomment-222107999>, or mute
> the thread
> <https://github.com/notifications/unsubscribe/AAvgn0V3BGiI-t5WE78tvlhmBgzF0WpHks5qFsAQgaJpZM4HB6m->
> .
>

gchatelet commented:
Changed `MurmurHash` to `MurmurHash3`, fixed the documentation rebased + squashed history.

@aappleby good to know :) I'm curious now!

9il commented on std/digest/murmurhash.d:
```diff
@@ -0,0 +1,753 @@
+        import std.conv;
+
+        static assert(false,
+                "MurmurHash3(" ~ size.to!string ~ "," ~ opt.to!string ~ ") is not implemented");
```
Do not need to import `std.conv`: `size.stringof` and `opt.stringof` should work

wilzbach commented on std/digest/murmurhash.d:
```diff
@@ -0,0 +1,751 @@
+
+    testUnalignedHash!(MurmurHash3!32)();
+    testUnalignedHash!(MurmurHash3!(128, 32))();
+    testUnalignedHash!(MurmurHash3!(128, 64))();
```
do you know about static foreach in D?

```
import std.meta: AliasSeq:
alias seq = AliasSeq;
foreach (H; seq!(32, seq!(128, 32), seq!(128, 64)))
{
 ....
}
```

9il replied:
only 3 lines without static foreach

wilzbach commented on std/digest/murmurhash.d:
```diff
@@ -0,0 +1,751 @@
+all architectures.
+
+The current version is MurmurHash3, which yields a 32-bit or 128-bit hash value.
+The older MurmurHash 1 and 2 are currently not supported.
```
this sounds negative - why not entirely remove this sentence. After all - who cares about older versions?

wilzbach replied:
It just sounded like a "missing feature", it was the "currently"  that triggered my internal _this library_ is incomplete flag.
In any case - it's a tiny nitpick ;-)

wilzbach commented on std/digest/murmurhash.d:
```diff
@@ -0,0 +1,751 @@
+ + $(LI `ubyte[Element.sizeof] getBytes()`)
+ + )
+ +/
+struct MurmurHash3(uint size /* 32 or 128 */ , uint opt = size_t.sizeof == 8 ? 64 : 32)
```
how about a template constraint?
```
if (size == 32 || size == 128 && opt == 32 || opt = 64)
```

9il replied:
No

wilzbach commented on std/digest/murmurhash.d:
```diff
@@ -0,0 +1,751 @@
+$(LI $(D MurmurHash3!32) produces a 32-bit value and is optimized for 32-bit architectures)
+$(LI $(D MurmurHash3!(128, 32)) produces a 128-bit value and is optimized for 32-bit architectures)
+$(LI $(D MurmurHash3!(128, 64)) produces a 128-bit value and is optimized for 64-bit architectures)
+)
```
Maybe add something like by default `MurmurHash3!128` picks your current architecture

or group those two categories in the listing, e.g. like this:

- MurmurHash!32 produces a 32-bit value and is optimized for 32-bit architectures
- MurmuHash!128 produces a 128-bit value and selects your current architecture by default
   - MurmurHash!(128, 32) optimized for 32-bit architectures
   - MurmurHash!(128, 64) optimized for 64-bit architectures

We should make it clear that manually passing the second parameter is usually not needed and when one should.

wilzbach replied:
> produce different digest results.
 
You are right, but my point was that we select it as default in the struct below, so we should tell the user here about it too (or remove the default). It was more that the doc and the actual struct seemed out of sync.

wilzbach commented:
@gchatelet please ignore my nitpicks - if we want we can still improve docs afterwards.
So can we go with `murmurhash` or does it need to go through the `std.experimental` hop?

gchatelet commented:
@wilzbach about experimental: since the interface is already standardized I don't think it makes sense.
I'd just go for it and improve. This PR has been initiated 6 months ago, let's merge and iterate.
I don't see any show stopper any more.

9il commented:
Auto-merge toggled on

9il commented:
PR for Slice to use MurmurHash3 #4639
